### PR TITLE
Handle ToC before handling shortcodes.

### PR DESCRIPTION
**Pull request by targodan**
The generation of the ToC is now done before the shortcodes are
handled. This makes the use of ToC in shortcodes possible.

The test checks whether or not the ToC is successfully rendered when
used from within a shortcode.

This relates to issue #2433.

### CODE REVIEW

File: hugolib/hugo_sites_test.go [hunk 1]
```diff
@@ -659,6 +670,106 @@ func TestChangeDefaultLanguage(t *testing.T) {
...
+	cfg := BuildCfg{}
+
+	err := sites.Build(cfg)
+
+	if err != nil {
+		t.Fatalf("Failed to build sites: %s", err)
+	}
+
+	// Not checking the heading-ids because they are not generated by the ToC.
+	assertFileContentRegexp(t, "public/en/post/simple/index.html", true, tocPageSimpleExpected)
```
Comment (bep) [inline hugolib/hugo_sites_test.go]: You can turn off those funky heading IDs by setting plainIDAnchors = true or something -- it is set in another test, so search for it.

Comment (targodan) [inline hugolib/hugo_sites_test.go]: Will get to it in a minute. Should I leave the function `assertFileContentRegexp` in for future use or should I remove it so as to not have dead code in it?

Comment (bep) [inline hugolib/hugo_sites_test.go]: Just leave it be, it will be useful at some point.

Comment (targodan) [inline hugolib/hugo_sites_test.go]: Ok, I just tried adding `plainIDAnchors = true` to the `[blackfriday]` section of the config but it didn't change anything. According to the [documentation](https://gohugo.io/overview/configuration/) it is true by default and it isn't what we are looking for.

My idea was that my test should not break when/if the id-generation is altered. I have found a different sollution for that. My test is now using the `#heading {#id}` method of secifying ids.

**Comment by bep** [comment]
Thanks for this!

**Comment by ketan** [comment]
Interestingly, I'm looking to generate TOC after shortcodes. My requirement is that I have some shortcodes that include partials. Some of these partials contain headers, which are not known being rendered by the toc.

Any thoughts on working around this issue?

**Comment by labkode** [comment]
Same problem as @ketan, I use a shortcode that generates configuration directives and they are not rendered into the TOC because they live in a shortcode.

**Comment by plinss** [comment]
Also similar problem as @ketan, I'm trying to use shortcodes in the headings to create icons, and the TOC now contains `HAHAHUGOSHORTCODE`.

**Pull request merged**